tests/integration/test_auth.py

```diff
@@ -0,0 +1,83 @@
+import pytest
+import responses
+from django.urls import reverse
+from furl import furl
+
+from jobserver.models import User
+
+
+@pytest.mark.django_db
+def test_login_pipeline(client):
+    assert User.objects.count() == 0
+
+    with responses.RequestsMock() as rsps:
+        access_token_url = "https://github.com/login/oauth/access_token"
+        access_token_data = {
+            "access_token": "dummy-token",
+            "token_type": "bearer",
+            "scope": "user:email",
+        }
+        rsps.add(responses.POST, access_token_url, json=access_token_data, status=200)
+
+        user_url = "https://api.github.com/user"
+        user_data = {
+            "login": "dummy-user",
+            "name": "Test User",
+            "email": None,
+        }
+        rsps.add(responses.GET, user_url, json=user_data, status=200)
+
+        emails_url = "https://api.github.com/user/emails"
+        emails_data = [
+            {
+                "email": "test@example.com",
+                "primary": True,
+                "verified": True,
+                "visibility": "private",
+            },
+        ]
+        rsps.add(responses.GET, emails_url, json=emails_data, status=200)
+
+        membership_url = "https://api.github.com/orgs/opensafely/members/dummy-user"
+        rsps.add(responses.GET, membership_url, status=204)
+
+        # set a dummy state value in the test Client's session to match the
+        # value in redirect_url below
+        session = client.session
+        session["github_state"] = "test-state"
+        session.save()
+
+        redirect_url = (
+            reverse("social:complete", kwargs={"backend": "github"})
+            + "?code=test-code&state=test-state"
+        )
+
+        response = client.get(redirect_url, secure=True)
+
+    assert response.status_code == 302
+
+    user = User.objects.first()
+    assert user.first_name == "Test"
+    assert user.last_name == "User"
+    assert user.email == "test@example.com"
+    assert user.notifications_email == "test@example.com"
+    assert user.is_active
+    assert not user.is_staff
+    assert not user.is_superuser
+
+
+@pytest.mark.django_db
+def test_login_redirects_correctly(client):
+    login_url = reverse("social:begin", kwargs={"backend": "github"})
+    response = client.get(login_url)
+
+    assert response.status_code == 302
+
+    f = furl(response.url)
+    assert f.scheme == "https"
+    assert f.host == "github.com"
+    assert str(f.path) == "/login/oauth/authorize"
+    assert f.args["client_id"] == "test"  # configured in pyproject.toml
+    assert "state" in f.args  # dynamically generated by the Auth Backend
+    assert f.args["redirect_uri"] == "http://testserver/complete/github/"
```

Comment:
Where does `testserver` come from?

Reply:
It's the hostname the testing server runs with, so if you were to call `request.build_absolute_uri()` inside a test/view under test that's what comes out.  I can't actually find anything in the docs which says _why_, but it's always been that way!